sklearn/decomposition/_base.py

```diff
@@ -60,8 +60,12 @@ def get_precision(self):
         # handle corner cases first
         if self.n_components_ == 0:
             return np.eye(n_features) / self.noise_variance_
-        if self.n_components_ == n_features:
-            return linalg.inv(self.get_covariance())
+        if hasattr(self, 'n_samples_'):
+            if self.n_components_ == min(self.n_samples_, n_features):
+                return linalg.inv(self.get_covariance())
+        else:
+            if self.n_components == n_features:
+                return linalg.inv(self.get_covariance())
```

Comment:
The `if n_components_ == n_samples/n_features` checks here seem like duplicates of the checks inside `IncrementalPCA` and `PCA`, see below:

https://github.com/scikit-learn/scikit-learn/blob/1ac9d6817da4d54d07f11e538d3447290cedfca3/sklearn/decomposition/_incremental_pca.py#L310-L314

https://github.com/scikit-learn/scikit-learn/blob/1ac9d6817da4d54d07f11e538d3447290cedfca3/sklearn/decomposition/_pca.py#L481-L484

Since noise variance is the root problem here, could we check `if self.noise_variance_ == 0.` instead? That way we wouldn't have to infer the PCA type using a `hasattr` trick.